src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets

```diff
@@ -276,9 +276,8 @@ Copyright (c) .NET Foundation. All rights reserved.
     -->
   <UsingTask TaskName="CreateAppHost" AssemblyFile="$(MicrosoftNETBuildTasksAssembly)" />
   <Target Name="_ComputeNETCoreBuildOutputFiles"
-          DependsOnTargets="ResolvePackageAssets"
-          AfterTargets="ResolveReferences"
-          BeforeTargets="AssignTargetPaths"
+          DependsOnTargets="ResolvePackageAssets;CoreCompile"
```

Comment:
>CoreCompile [](start = 49, length = 11)

I'm not sure about this dependency on CoreCompile. I know we need the intermediate assembly available, but there could be scenarios where we end up triggering a re-build when we were supposed to be using what is already on disk. This has come up a few times with other targets. I have to think about the best approach here.

Reply:
Still thinking about it, but I thing following the existing idioms in common targets would be to generate the apphost to intermediate directory upon/after compile and then designate the intermediate item as something to copy separately.

---
In reply to: [219644421](https://github.com/dotnet/sdk/pull/2545#discussion_r219644421) [](ancestors = 219644421)

Reply:
@nguerrera generating the apphost into the intermediate directory would still have a dependency on CoreCompile, wouldn't it?

Reply:
Correct. 

I was thinking about changing the paradigm to react to compilation instead of depending on compilation. 

I am concerned that there may be places where this target is invoked without expecting compilation to happen even if the source files are out of date. Look at the long discussion on https://github.com/dotnet/cli/issues/5331 for some background on why this is worth worrying about.

What I'm wondering is: can the apphost generation happen in a separate target that is AfterTargets=CoreCompile (after we compile a managed exe to the intermediate directory, create the matching apphost next to it). And this target would just add the intermediate location to the set of files to be copied.

Adding @rainersigwald who may have other ideas here.

Reply:
Agreed that if the apphost creation will be invoked in a way that doesn't expect compilation to happen, it shouldn't depend on CoreCompile. But the current target logically depends on an up-to-date IntermediateAssembly, and I think "DependsOnTargets=CoreCompile" captures that.

If the scenario you're thinking about is one where we do "publish --no-build", and expect that this will create an apphost for the previously-built intermediate assembly (even if source files are out of date), then I can see your concern. To me it makes sense to implement this by expressing the dependency on the outputs of a target that logically produces IntermediateAssembly, and having that target internally respect the --no-build option by using the existing IntermediateAssembly in that case. This way the --no-build behavior is transparent to targets that depend on build outputs.

AfterTargets=CoreCompile could work, but that has other issues. Any inputs to the apphost generator that don't come from CoreCompile should trigger apphost re-generation if they change. For example, setting the target to WinExe should cause the apphost to be re-generated with the gui bit, even if we don't want to re-compile.

Reply:
> If the scenario you're thinking about is one where we do "publish" --no-build

I'm thinking even more generally. This will cause any caller of GetCopyToOutputDirectoryItems to trigger compilation and I don't think that's OK. 

Agree that AfterTargets="CoreCompile" alone doesn't solve it. I'm still thinking about what's best here.




---
In reply to: [220379741](https://github.com/dotnet/sdk/pull/2545#discussion_r220379741) [](ancestors = 220379741)

Reply:
It generally makes sense for GetCopyToOutputDirectoryItems to trigger compilation, because it logically depends on the compilation output through the apphost. If we want to optimize for "publish --no-build", we need to cut out some some dependency edges from Publish -> CoreCompile in this case (or alternatively make CoreCompile do nothing).

In https://github.com/dotnet/sdk/pull/2111, it looks like we already cut one edge from Publish -> CoreCompile by making Publish not depend on Build in the "--no-build" case. Maybe we want to do something similar here? We could make GetCopyToOutputDirectoryItems depend on _ComputeNETCoreBuildOutputFiles unless NoBuild is specified. If NoBuild is specified, GetCopyToOutputDirectoryItems could just try to get the last-built apphost from the intermediate directory.

Just an idea, but I'm sure there are more constraints I don't know about. I'll let you decide how you think this should be done - let me know if there's anything I can do to help!

Reply:
`GetCopyToOutputDirectoryItems` should _not_ do actual work and should not depend on `CoreCompile` (and friends), because it is called on references in the `/p:BuildProjectReferences=false` case, which should not trigger compilation in the referenced projects. I don't think the `NoBuild` case applies here for that reason--you're building, but some other project.

Unfortunately, I can't think of a great way to accomplish this. Best I can think of is to move the actual work into a target (`CreateAppHost`?) and append `;CreateAppHost` to `$(CompileDependsOn)`, which would insert it in the normal flow shortly after `CoreCompile` but not require it to express the dependency. Then, do no real work in `_ComputeNETCoreBuildOutputFiles` and run it `BeforeTargets="GetCopyToOutputDirectoryItems"`.

Reply:
Thanks, @rainersigwald. This sounds like a good approach. @sbomer, can you look at doing it that way?

Reply:
Thanks @rainersigwald and @nguerrera. I think the latest change does the trick - PTAL.